Mathlib/Data/PNat/Xgcd.lean

```diff
@@ -0,0 +1,592 @@
+  ⟨fun u => u.bp⟩
+
+/-- The `Repr` instance converts terms to strings in a way that
+ reflects the matrix/vector interpretation as above. -/
+instance : Repr XgcdType where
+  reprPrec
+  | g, _ => s!"[[[ {repr (g.wp + 1)}, {(repr g.x)} ], [" ++
+            s!"{repr g.y}, {repr (g.zp + 1)}]], [" ++
+            s!"{repr (g.ap + 1)}, {repr (g.bp + 1)}]]"
+
+/-- Another `mk` using ℕ and ℕ+ -/
+def mk' (w : ℕ+) (x : ℕ) (y : ℕ) (z : ℕ+) (a : ℕ+) (b : ℕ+) : XgcdType :=
+  mk w.val.pred x y z.val.pred a.val.pred b.val.pred
+#align pnat.xgcd_type.mk' PNat.XgcdType.mk'
+
+/-- `w = wp + 1` -/
+def w : ℕ+ :=
+  succPNat u.wp
+#align pnat.xgcd_type.w PNat.XgcdType.w
+
+/-- `z = zp + 1` -/
+def z : ℕ+ :=
+  succPNat u.zp
+#align pnat.xgcd_type.z PNat.XgcdType.z
+
+/-- `a = ap + 1` -/
+def a : ℕ+ :=
+  succPNat u.ap
+#align pnat.xgcd_type.a PNat.XgcdType.a
+
+/-- `b = bp + 1` -/
+def b : ℕ+ :=
+  succPNat u.bp
+#align pnat.xgcd_type.b PNat.XgcdType.b
+
+/-- `r = a % b`: remainder -/
+def r : ℕ :=
+  (u.ap + 1) % (u.bp + 1)
+#align pnat.xgcd_type.r PNat.XgcdType.r
+
+/-- `q = ap / bp`: quotient -/
+def q : ℕ :=
+  (u.ap + 1) / (u.bp + 1)
+#align pnat.xgcd_type.q PNat.XgcdType.q
+
+/-- `qp = q - 1` -/
+def qp : ℕ :=
+  u.q - 1
+#align pnat.xgcd_type.qp PNat.XgcdType.qp
+
+/-- The map `v` gives the product of the matrix
+ [[w, x], [y, z]] = [[wp + 1, x], [y, zp + 1]]
+ and the vector [a, b] = [ap + 1, bp + 1].  The map
+ `vp` gives [sp, tp] such that v = [sp + 1, tp + 1].
+-/
+def vp : ℕ × ℕ :=
+  ⟨u.wp + u.x + u.ap + u.wp * u.ap + u.x * u.bp, u.y + u.zp + u.bp + u.y * u.ap + u.zp * u.bp⟩
+#align pnat.xgcd_type.vp PNat.XgcdType.vp
+
+/-- `v = [sp + 1, tp + 1]`, check `vp` -/
+def v : ℕ × ℕ :=
+  ⟨u.w * u.a + u.x * u.b, u.y * u.a + u.z * u.b⟩
+#align pnat.xgcd_type.v PNat.XgcdType.v
+
+/-- `succ₂ [t.1, t.2] = [t.1.succ, t.2.succ]` -/
+def succ₂ (t : ℕ × ℕ) : ℕ × ℕ :=
+  ⟨t.1.succ, t.2.succ⟩
+#align pnat.xgcd_type.succ₂ PNat.XgcdType.succ₂
+
+theorem v_eq_succ_vp : u.v = succ₂ u.vp := by
+  ext <;> dsimp [v, vp, w, z, a, b, succ₂] <;> (repeat' rw [Nat.succ_eq_add_one]; ring_nf)
+#align pnat.xgcd_type.v_eq_succ_vp PNat.XgcdType.v_eq_succ_vp
+
+/-- `is_special` holds if the matrix has determinant one. -/
+def is_special : Prop :=
+  u.wp + u.zp + u.wp * u.zp = u.x * u.y
+#align pnat.xgcd_type.is_special PNat.XgcdType.is_special
+
+/-- `is_special'` is an alternative of `is_special`. -/
+def is_special' : Prop :=
+  u.w * u.z = succPNat (u.x * u.y)
+#align pnat.xgcd_type.is_special' PNat.XgcdType.is_special'
+
+theorem is_special_iff : u.is_special ↔ u.is_special' := by
+  dsimp [is_special, is_special']
+  let ⟨wp, x, y, zp, ap, bp⟩ := u
+  constructor <;> intro h <;> simp [w, z, succPNat] at * <;>
+    simp only [← coe_inj, mul_coe, mk_coe] at *
+  . simp_all [← h, Nat.mul, Nat.succ_eq_add_one]; ring
+  . simp [Nat.succ_eq_add_one, Nat.mul_add, Nat.add_mul, ← Nat.add_assoc] at h; rw [← h]; ring
+  -- Porting note: Below doesn't work, so I fixed it as above.
```

Comment:
I think you can delete the entire commented code below.
Your new proof is so much shorter that I don't think anybody is interested in reviving the old proof.